opentelemetry-sdk/src/opentelemetry/sdk/metrics/view.py

```diff
@@ -0,0 +1,107 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import threading
+
+logger = logging.getLogger(__name__)
+
+
+class ViewData:
+
+    def __init__(self, labels, aggregator_type):
+        self.labels = labels
+        self.aggregator = aggregator_type()
+
+
+class View:
+
+    def __init__(self, metric, aggregator_type, label_keys=None):
+        self.metric = metric
+        self.aggregator_type = aggregator_type
+        self._view_datas = {} # Labels to ViewData
+        self._view_datas_lock = threading.Lock()
+        # TODO: add configuration for aggregators (histogram, etc.)
+        # TODO: add label key configuration
+        self.label_keys = label_keys
+        if self.label_keys is None:
+            self.label_keys = []
+
+    def get_all_view_data(self):
+        # We return a copy of the values because the dictionary can be can be
+        # always changing
+        return list(self._view_datas.values())
+
+    def get_view_data(self, labels):
+        return self._view_datas.get(labels)
+
+    def update_view_datas(self, labels, value):
+        with self._view_datas_lock:
+            view_data = self.get_view_data(labels)
+            if view_data is None:
+                view_data = ViewData(labels, self.aggregator_type)
+                self._view_datas[labels] = view_data
+            view_data.aggregator.update(value)
+
+    # Uniqueness of View is based on metric and aggregation type
+    def __hash__(self):
+        return hash((self.metric, self.aggregator_type))
+
+    def __eq__(self, other):
+        return self.metric == other.metric and \
+            self.aggregator_type == other.aggregator_type
```

Comment:
Just reading the [internal details](https://github.com/open-telemetry/oteps/blob/46948ac8f175c95683894dd9656d2980ce4208a1/text/0089-metrics-views-api.md#internal-details) section of the otep, it sounds like the uniqueness of a view is derived solely on the aggregation type and metric instrument.

I think its less about views with different label keys being equal, more as views with different labels keys "can't" be equal. Consider the following scenario:

Metric m1 = Counter()
View a = View(m1, CountAggregation, keys=["asd, def"])
View b = View(m1, CountAggregation, keys=["def"])

m1.record(100, labels={"def":"value"})

If both views were to be configured to drop keys, {"def":"value"} would still be recorded since the key exists in both. However, which aggregation does the record belong to? If we choose to treat Views with different label keys differently, we would have to compute the "final set of label keys to aggregate by" between every view and then create aggregators based off how many unique labels are generated (since records with the same labels are aggregated  together). This is also poses a problem of timing, where what if a view is defined after a collection pass?

Metric m1 = Counter()
View a = View(m1, CountAggregation, keys=["asd, def"])

m1.record(100, labels={"def":"value"})

// collect happens here //

View b = View(m1, CountAggregation, keys=["def"])

m1.record(200, labels={"def":"value"})

By the time the second record is hit, an aggregation already exists for {"def":"value"}. But ON RECORD of the second record, we do not do the "dropping of the label keys logic to find the unique labelset to aggregate on" logic YET, so we do not know that it is supposed to use the same aggregator.

Thoughts?

Reply:
> However, which aggregation does the record belong to?

We might be thinking about this differently. I'd expect both aggregations to be updated.

For view `a`, update the aggregator (or aggregation data) associated with:

```
{"asd": None, "def": "value"}
```

and for view `b`:
```
{"def": "value"}
```

I could imagine optimizing this later so we're only storing what's required to reconstruct the aggregated values at export time -- so in this case only the aggregated values for each label set for view `a` -- but for now I think we should just focus on getting the API right.

As for adding a new view in the middle of an export interval, maybe we should just require the user to define all their views before we start collecting metrics.

Reply:
I was referring more to the case where the final resulting labels for an aggregator is generated by two different views. A simple example would be the one above, but they both are configured to keep all their labels. Thus any record would iterate through both views, but the ending labels would only be one unique set. This would mean the aggregation data itself needs to be decoupled from the View layer.

Reply:
Although I'm thinking this case might be solved by [your comment](https://github.com/open-telemetry/opentelemetry-python/pull/596#discussion_r411767186) here.

Reply:
Actually this should be solved by [your comment](https://github.com/open-telemetry/opentelemetry-python/pull/596#discussion_r411767186)

Reply:
> As for adding a new view in the middle of an export interval, maybe we should just require the user to define all their views before we start collecting metrics.

I think this will help with automatically using default aggregations if views are not created.